include/execution.hpp

```diff
@@ -1873,6 +1860,251 @@ namespace std::execution {
   struct sender_traits<__on::__impl::__sender<SchedulerId, SenderId>>
     : sender_traits<__t<SenderId>> {};
 
+  /////////////////////////////////////////////////////////////////////////////
+  // [execution.senders.adaptors.when_all]
+  inline namespace __when_all {
+    namespace __impl {
+      enum __state_t { __started, __error, __done };
+
+      struct __on_stop_requested {
+        in_place_stop_source& __stop_source_;
+        void operator()() noexcept {
+          __stop_source_.request_stop();
+        }
+      };
+
+      template <class... SenderIds>
+        struct __sender {
+          template <template <class...> class Tuple,
+                    template <class...> class Variant>
+            using value_types = Variant<__meta_invoke<__concat<Tuple>,
+                typename sender_traits<__t<SenderIds>>::template
+                    value_types<__types, __single_t>...>>;
+
+          template <template <class...> class Variant>
+            using error_types =
+                __meta_invoke<__concat<__unique<Variant>::template __f>,
+                    __types<exception_ptr>,
+                    typename sender_traits<__t<SenderIds>>::template
+                        error_types<__types>...>;
+
+          static constexpr bool sends_done = true;
+
+          template <class... Sndrs>
+            explicit __sender(Sndrs&&... sndrs)
+              : __sndrs_((Sndrs&&) sndrs...)
+            {}
+
+         private:
+          template <class CvrefReceiverId>
+            struct __operation {
+              using WhenAll = __member_t<CvrefReceiverId, __sender>;
+              using Receiver = __t<decay_t<CvrefReceiverId>>;
+              template <size_t Index>
+                struct __receiver : receiver_adaptor<__receiver<Index>> {
+                  Receiver&& base() && noexcept {
+                    return (Receiver&&) __op_->__recvr_;
+                  }
+                  const Receiver& base() const & noexcept {
+                    return __op_->__recvr_;
+                  }
+                  template <class... Values>
+                    void set_value(Values&&... vals) && noexcept try {
+                      __state_t expected = __started;
+                      // We only need to bother recording the completion values
+                      // if we're not already in the "error" or "done" state.
+                      if (__op_->__state_ == __started) {
+                        std::get<Index>(__op_->__values_).emplace(
+                            (Values&&) vals...);
+                      }
+                      (void) __op_->__barrier_.arrive();
+                    } catch(...) {
+                      ((__receiver&&) *this).set_error(current_exception());
+                    }
+                  template <class Error>
+                      requires receiver<Receiver, Error>
+                    void set_error(Error&& error, __state_t expected = __started) && noexcept {
```

Comment:
I would prefer an internal method that set_value, set_error and set_done can call with an explicit expected state.

Reply:
I'm not sure what you mean. One method for all three signals? But the code is different for each.

Reply:
The idea was to make an internal __set_error(Error, __state_t) and call that from set_value and set_done instead of this->set_error() with a defaulted arg. set_error() could forward to __set_error as well.